File: 02. Algorithms/09. Dynamic Programming/006. House Robber II/code.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size() ;
        if(n==1)
            return nums[0] ;
        if(n==2)
            return max(nums[0],nums[1]) ;
        
        int dp1[n-1] ;
        
        dp1[0] = nums[0] ;
        dp1[1] = max(nums[0],nums[1]) ;
        for(int i=2;i<n-1;i++) {
            dp1[i] = max(dp1[i-1],nums[i]+dp1[i-2]);
        }
        
        int dp2[n-1] ;
        
        dp2[0] = nums[1] ;
        dp2[1] = max(nums[2],nums[1]) ;
        for(int i=2;i<n-1;i++) {
            dp2[i] = max(dp2[i-1],nums[i+1]+dp2[i-2]);
        }
        
        return max(dp1[n-2],dp2[n-2]) ;
    }
};

File: 02. Algorithms/09. Dynamic Programming/006. House Robber II/code.js
/**
 * @param {number[]} nums
 * @return {number}
 */
    
    
var rob = function(nums) {
    
    if(nums.length === 1) return nums[0];
    
    let includeFirst = preRob(nums.slice(1));
    let notIncludeFirst = preRob(nums.slice(0, nums.length - 1));
    return Math.max(includeFirst, notIncludeFirst);
    
};

var preRob = function(nums) {
    
    let rob1 = 0;
    let rob2 = 0;
    
    let temp = 0;
    for(let i = 0; i < nums.length; i++) {
        temp = Math.max(rob1 + nums[i], rob2);
        rob1 = rob2;
        rob2 = temp;
    }
    
    
    return rob2;
}
